File: trunk/src/cpp/items/character.cpp
#include "engines/log.hpp"
#include "engines/mediator.hpp"
#include "items/character.hpp"
#include "map.hpp"
#include "messages/nothing.hpp"
#include "movable_model.hpp"
#include "pair.hpp"
#include "set.hpp"
#include "skill.hpp"
#include "slot.hpp"
#include "strategies/behaviors/ai.hpp"
#include "strategies/behaviors/player.hpp"

using namespace std;
using namespace Radakan;
using namespace Radakan :: Items;

//	static
string Character ::
	get_class_name ()
{
	return "Items :: Character";
}

//	static
Reference <Character> Character ::
	get_player_character ()
{
	return Strategies :: Behaviors :: Player :: get () -> character;
}

//  constructor
Character ::
	Character
	(
		string new_name,
		float new_mass,
		Mathematics :: Vector_3D new_size,
		const Reference <Mesh_Data> new_mesh_data,
		string type
	) :
	Object (new_name),
	Container_Item <Item>
	(
		new_name,
		new_mass,
		new_size,
		new_mesh_data
	),
	//	head (new Static_Item (name + "'s head", "bar.mesh", 1, 1)),
	//	head (new Container_Item <Hat> ()),
	//	body (new Container_Item <Shirt> ()),
	back
	(
		new Container_Item <Container_Item <Item> >
		(
			name + "'s back",
			1,
			Mathematics :: Vector_3D (0.5, 0.5, 0.3),
			Reference <Mesh_Data> ()
		)
	),
	//	arms (new Container_Item <Bracer> ()),
	right_hand
	(
		new Container_Item <Item>
		(
			name + "'s right hand",
			1,
			Mathematics :: Vector_3D (1, 0.3, 0.3),
			Reference <Mesh_Data> ()
		)
	),
	//	legs (new Container_Item <Pants> ()),
	//	feet (new Container_Item <Shoe> ()),
	skills (new Map <string, Skill> ("skills"))
{
	Engines :: Log :: trace (me, Character :: get_class_name (), "", to_string (new_mass),
		new_size . to_string (), new_mesh_data . get_name ());
	add_automatic_destruction_prevention ("construction of " + get_class_name ());
	
	bool check/* = Container_Item <Item> :: add (head)*/;
	/*assert (check);*/
	//	Container_Item <Item> :: add (body);
	check = Container_Item <Item> :: add (back);
	assert (check);
	//	Container_Item <Item> :: add (arms);
	check = Container_Item <Item> :: add (right_hand);
	assert (check);
	//	Container_Item <Item> :: add (legs);
	//	Container_Item <Item> :: add (feet);

	//	Make sure no body parts are added anymore.
	Container <Items :: Item> :: seal ();

	if (type == "player character")
	{
		behave <Strategies :: Behaviors :: Player> ();
	}
	else
	{
		assert (type == "npc");

		behave <Strategies :: Behaviors :: AI> ();
	}

	Engines :: Mediator :: get () -> register_observer <Messages :: Nothing>
		(Reference <Observer <Messages :: Nothing> > (this));

	remove_automatic_destruction_prevention ("construction of " + get_class_name ());
	assert (is_initialized ());
}

//  destructor
Character ::
	~Character ()
{
	Engines :: Log :: trace (me, Character :: get_class_name (), "~");
	assert (Character :: is_initialized ());

	//	Do nothing.
	//	'prepare_for_destruction ();' hasn't to be called,
	//	because this is an abstract base class.
}

//	virtual
bool Character ::
	is_initialized ()
	const
{
	assert (Observer <Messages :: Nothing> :: is_initialized ());
	assert (Container_Item <Item> :: is_initialized ());

	return true;
}

void Character ::
	drop (Reference <Strategies :: Behaviors :: Behavior> behavior)
{
	get_movable_model () -> turn (1, get_model () -> get_side_direction ());
	
	Engines :: Log :: show (me . get_name (true) + " died.");

	Slot <Strategies :: Behaviors :: Behavior> :: drop (behavior);
}

bool Character ::
	is_alive () const
{
	return has_active_state ();
}

Reference <Movable_Model> Character ::
	get_movable_model ()
	const
{
	if (! movable_model . points_to_object ())
	{
		movable_model . reset_pointee (get_model () . cast <Movable_Model> (), true);
	}

	return movable_model;
}

void Character ::
	call (Reference <Messages :: Nothing>)
{
	assert (is_initialized ());

	run (Messages :: Nothing :: get ());
}

float Character ::
	get_skill (const string & skill_name)
	const
{
	assert (is_initialized ());
	assert (skill_name != "fear");
	assert (skill_name != "like");
	
	if (! skills -> look_up (skill_name) . points_to_object ())
	{
		skills -> add
		(
			Pair <string, Skill> :: create
				(skill_name, Reference <Skill> (new Skill (skill_name)))
		);
	}

	return skills -> look_up (skill_name) -> get_value ();
}

template <class T> void Character ::
	behave ()
{
	assert (! State_Machine <Strategies :: Behaviors :: Behavior> :: has_active_state ());
	
	set_active_state (Reference <Strategies :: Behaviors :: Behavior>
		(new T (Reference <Character> (this))));
}

File: radakan/src/com/gibbon/radakan/ai/Character.java
/*
 * Radakan RPG is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Radakan RPG is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Radakan RPG.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.gibbon.radakan.ai;

import java.util.*;

///	All characters observe each other.
public final class Character extends ContainerItem <Item> implements EventListener <Nothing> {
	Character(String newName,float newMass, Vector3D newSize, MeshData newMeshData, String type) {
		super
		(
			newName,
			newMass,
			newSize,
			newMeshData
		);
		name = newName;

		
		Behavior behavior;
		if(type == "player character") {
			behavior = new Player(this);
		}
		else {
			behavior = new ArtificialIntelligence(this);
		}
		stateMachine = new StrategyStateMachine <Nothing, Behavior> (behavior);
		
		
		//	head(new StaticItem(name + "'s head", "bar.mesh", 1, 1)),
		//	head(new ContainerItem <Hat>()),
		//	body(new ContainerItem <Shirt>()),
		back
		(
			new ContainerItem <ContainerItem <Item> >
			(
				name + "'s back",
				1,
				Vector3D(0.5, 0.5, 0.3),
				MeshData()
			)
		);
		//	arms(new ContainerItem <Bracer>()),
		rightHand
		(
			new ContainerItem <Item>
			(
				name + "'s right hand",
				1,
				Mathematics.Vector3D(1, 0.3, 0.3),
				MeshData()
			)
		);
		//	legs(new ContainerItem <Pants>()),
		//	feet(new ContainerItem <Shoe>()),
		skills(new map <String, <Skill> >) (
			newSize.toString(), newMeshData.name);
		
		Boolean check/* = ContainerItem <Item>.insert(head)*/;
		//	ContainerItem <Item>.insert(body);
		check = ContainerItem <Item>.insert(back);
		//	ContainerItem <Item>.insert(arms);
		check = ContainerItem <Item>.insert(rightHand);
		//	ContainerItem <Item>.insert(legs);
		//	ContainerItem <Item>.insert(feet);
	
		//	Make sure no body parts are added anymore.
		seal();
	
		Mediator.nothingMediator.registerObserver(this);

	}

	public final StrategyStateMachine <Nothing, Behavior> stateMachine;
		
	public static Character getPlayerCharacter() {
		return Player.get().character;
	}
	
	static String Arachnyd;
	static String Grogg;
	static String Harpy;
	static String Inferno;
	static String Xemna;

	public void call(Nothing message) {
		stateMachine.run(Nothing.get());
	}

	public Boolean isAlive(){
		return stateMachine.hasState();
	}

	float getSkill(String skillName)	{

		Map <String, Skill>.iterator skill = skills.find(skillName);

		//	when the skill couldn't be found
		if(skill == skills.end()) {
			skill = skills.insert
			(
				pair <String, Skill>
					(skillName, new Skill(skillName))
			).first;
		}

		return skill.second.getValue();
	}
	
	public final String name;

	public String specie;

	//	public Item & head;
	//	public Multislot <Shirt> & body;
	public ContainerItem <ContainerItem <Item> > back;
	//	public Multislot <Bracer> & arms;
	public ContainerItem <Item> leftHand;
	public ContainerItem <Item> rightHand;
	//	public Multislot <Pants> & legs;
	//	public Multislot <Shoe> & feet;

	public Map<String, Skill> skills = new HashMap<String, Skill>();
};
